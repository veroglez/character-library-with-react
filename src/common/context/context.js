import { node } from 'prop-types';
import React, { createContext, Component } from 'react';
import { C, L10N } from '..';

const Context = createContext();
const { Provider, Consumer: ConsumerData } = Context;

class ProviderData extends Component {
  static propTypes = {
    children: node,
  };

  static defaultProps = {
    children: undefined,
  };

  state = {
    next: '',
    characters: [],
    l10n: L10N,
  }

  componentDidMount() {
    const service = 'character';
    this.onFetch(service);
  }

  onFetch = async (service, next) => {
    const { characters } = this.state;
    const response = await fetch(next || `${C.DOMAIN}${service}`);
    const character = await response.json();
    this.setState({ characters: next ? [...characters, ...character.results] : character.results, next: character.info.next });
  }

  onData = object => this.setState(object);

  render() {
    const {
      onData, onFetch, props: { children },
    } = this;

    return (
      <Provider value={{ ...this.state, onData, onFetch }}>
        { children }
      </Provider>
    );
  }
}

export { ConsumerData, ProviderData };
