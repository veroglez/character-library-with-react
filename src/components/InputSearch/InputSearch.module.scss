@import "../../common/style.scss";

.input {
  height: 30px;
  border-radius: 5px;
  border: 1px solid map-get($color, light );
  width: 30%;
  padding: 2px 10px;
}