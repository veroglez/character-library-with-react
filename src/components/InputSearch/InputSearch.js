import React, { PureComponent, Fragment } from 'react';
import scss from './InputSearch.module.scss';
import { ConsumerData } from '../../common/context/context';

class InputSearch extends PureComponent {
  state = {
    value: '',
    busy: true,
    characters: [],
  }

  onState = ({ characters }) => this.setState({ busy: false, characters });

  onChange = (event, { onData }) => {
    const { characters } = this.state;
    this.setState({ value: event.target.value });

    const charactersFiltered = characters.filter(character => (
      character.name.toLowerCase().includes(event.target.value)
    ));

    onData({ characters: charactersFiltered });
  }

  render() {
    const { onChange, onState, state: { busy, value } } = this;

    return (
      <ConsumerData>
        {context => (
          <Fragment>
            {busy && context.characters.length > 0 && onState(context)}
            <input
              className={scss.input}
              placeholder={context.l10n.SEARCH}
              type="text"
              value={value}
              name="name"
              onChange={event => onChange(event, context)}
            />
          </Fragment>
        )}
      </ConsumerData>
    );
  }
}

export default InputSearch;
