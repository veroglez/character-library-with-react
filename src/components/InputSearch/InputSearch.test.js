import React from 'react';
import renderer from 'react-test-renderer';
import InputSearch from './InputSearch';
import { ProviderData } from '../../common/context/context';

const HOC = inherit => (
  <ProviderData l10n={{ SEARCH: 'Search' }}>
    <InputSearch {...inherit} />
  </ProviderData>
);

describe('<InputSearch>', () => {
  it('renders', () => {
    const tree = renderer.create(<HOC />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
