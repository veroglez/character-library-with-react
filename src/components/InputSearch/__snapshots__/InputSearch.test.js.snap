// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputSearch> renders 1`] = `
<input
  name="name"
  onChange={[Function]}
  placeholder="Search..."
  type="text"
  value=""
/>
`;
